src/main/scala/com/spark3d/geometry/ShellEnvelope.scala

```diff
@@ -0,0 +1,262 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.spark3d.geometry
+
+import com.spark3d.geometryObjects._
+
+/** Defines a shell of 3D coordinate space. Shell here is made by a difference of
+  * of two two concentric spheres. This can be used to define a bounding box of a geometryObject
+  *
+  * An Shell Envelope can be uniquely defined based on the center and its outer and inner radius.
+  *
+  * @param center coordinates of the center of the shell Envelope
+  * @param innerRadius inner radius of the shell Envelope
+  * @param outerRadius outer radius of the shell Envelope
+  */
+class ShellEnvelope(
+    var center: Point3D,
+    var innerRadius: Double,
+    var outerRadius: Double)
+  extends Envelope {
+
+  /**
+    * Creates a null shell Envelope.
+    */
+  def this() {
+    this(new Point3D(0.0, 0.0, 0.0), -1.0, -2.0)
+  }
+
+
+  /**
+    * Creates a shell Envelope defined with a center coordinates, inner and outer radius.
+    *
+    * @param x x-coordinate of the center of the sphere Envelope
+    * @param y y-coordinate of the center of the sphere Envelope
+    * @param z z-coordinate of the center of the sphere Envelope
+    * @param innerRadius inner radius of the Envelope
+    * @param outerRadius outer radius of the Envelope
+    */
+  def this(x: Double, y: Double, z: Double, innerRadius: Double, outerRadius: Double) {
+    this(new Point3D(x, y, z), innerRadius, outerRadius)
+  }
+
+  /**
+    * Clones the existing shell Envelope
+    *
+    * @param env shell Envelope to be cloned
+    */
+  def this(env: ShellEnvelope) {
+    this(env.center, env.innerRadius, env.outerRadius)
+  }
+
+  /**
+    * Returns if this is a null shell envelope.
+    *
+    * @return if this shell Envelope is null (empty Geometry) or not
+    */
+  def isNull(): Boolean = {
+    return if (innerRadius < 0 || outerRadius < 0 || innerRadius > outerRadius) true else false
+  }
+
+  /**
+    * Sets this shell Envelope to a null.
+    *
+    */
+  def setToNull: Unit = {
+    innerRadius = -1
+    outerRadius = -1
+  }
+
+  /**
+    * Returns the area of the shell Envelope.
+    *
+    * @return the area of the shell envelope, 0.0 if it is null
+    */
+  def getArea(): Double = {
+
+    if (isNull) {
+      return 0.0
+    }
+
+    val pi = 3.14159
+    4 * pi * ((outerRadius * outerRadius) - (innerRadius * innerRadius))
+  }
+
+  /**
+    * Expands the shell Envelope so that it contains the given Point. This will expand both the inner and outer
+    * radius of the shell Envelope by same amount.
+    *
+    * @param p the Point to expand to include
+    */
+  def expandToInclude(p: Point3D): Unit = {
```

Comment:
Question: I understand you expand both side to include the point.
But what about the following situation:
center (0, 0, 0)
innerRadius = 1
OuterRadius = 2
p (10, 10, 10)

You will increase both sides by sqrt(300), that is innerRadius will become negative, and your shell will become null?

To avoid this situation, it would be better to expand only the side where the point is?

Reply:
yes, makes sense, expanding only the side where the point is a better option. Will make this change.